### PR TITLE
Reorganisation and implementation of token refresh

### PR DESCRIPTION
**Pull request by enolfc**
Brings `egiauthenticator` and `egispawner` code here, so will deprecate existing repos.
Refreshes access tokens and makes then available via secrets to the notebook

### CODE REVIEW

**Comment by enolfc** [comment]
Still testing is missing and the egi-hub-addons are not yet migrated here

**Comment by enolfc** [comment]
To add to the deployment of notebooks:
```
kind: Role
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: hub-secrets
rules:
  - apiGroups: [""]       # "" indicates the core API group
    resources: ["secrets"]
    verbs: ["get", "watch", "list", "create", "delete", "patch"]
---
kind: RoleBinding
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: hub-secrets
subjects:
  - kind: ServiceAccount
    name: hub
    namespace: dev-hub
roleRef:
  kind: Role
  name: hub-secrets
  apiGroup: rbac.authorization.k8s.io
```

File: Dockerfile [hunk 1]
```diff
@@ -2,29 +2,25 @@
...
+COPY . /egi-notebooks-hub/
 
-# install the EGI Spawner
-RUN pip3 install git+https://github.com/enolfc/egispawner
+# install the hub extensions
+RUN pip3 install /egi-notebooks-hub/
 
-RUN pip3 install git+https://github.com/enolfc/egi-hub-addons
+# install d4science auth plugin for using the service in D4Science
+RUN pip3 install git+https://github.com/enolfc/d4scienceauth.git
```
Comment (andrea-manzi) [inline Dockerfile:13]:
```suggestion
RUN pip3 install git+https://github.com/enolfc/egi-hub-addons
```

Comment (andrea-manzi) [reply Dockerfile:13]: I put it back for now..

File: Dockerfile [hunk 2]
```diff
@@ -2,29 +2,25 @@
...
-COPY 401.html /usr/local/share/jupyterhub/templates/403.html
-
-COPY egi-notebooks.png /usr/local/share/jupyterhub/static/images/jupyter.png
-COPY notebooks-logo.png /usr/local/share/jupyterhub/static/images/notebooks-logo.png
-COPY cesnet.png /usr/local/share/jupyterhub/static/images/cesnet.png
-COPY cesga.jpg /usr/local/share/jupyterhub/static/images/cesga.jpg
-COPY infn-ct.jpg /usr/local/share/jupyterhub/static/images/infn-ct.jpg
+RUN cp /egi-notebooks-hub/templates/login.html /usr/local/share/jupyterhub/templates/login.html \
+    && cp /egi-notebooks-hub/templates/401.html /usr/local/share/jupyterhub/templates/401.html \
+    && cp /egi-notebooks-hub/templates/401.html /usr/local/share/jupyterhub/templates/403.html \
```
Comment (andrea-manzi) [inline Dockerfile:18]:
```suggestion
    && cp /egi-notebooks-hub/templates/403.html /usr/local/share/jupyterhub/templates/403.html \
```

Comment (enolfc) [reply Dockerfile:18]: This is intended like that, we don't have a 403 template

File: Dockerfile (hunk 1 above)
Comment (enolfc) [reply Dockerfile:13]: I have added the code in a push below, needs to be tested, but should mostly work

File: Dockerfile (hunk 2 above)
Comment (andrea-manzi) [reply Dockerfile:18]: ops, sorry:-)

File: egi_notebooks_hub/egispawner.py [hunk 3]
```diff
@@ -90,3 +92,87 @@ def set_access_token(self, token):
...
+                        "root": datahub_response["name"],
+                        "class": "onedatafs_jupyter.OnedataFSContentsManager",
+                        "config": {"space": "/" + datahub_response["name"] },
+                    })
+                except HTTPError as e:
+                    self.log.info("Something failed! %s", e)
+                    raise e
+            pod.spec.containers[0].args = (pod.spec.containers[0].args +
+                [
+                    '--NotebookApp.contents_manager_class=%s' % manager_class,
```
Comment (andrea-manzi) [inline egi_notebooks_hub/egispawner.py:169]:
```suggestion
                    '--NotebookApp.contents_manager_class=%s' % self.manager_class,
```

Comment (andrea-manzi) [reply egi_notebooks_hub/egispawner.py:169]: looks like self is missing here

**Pull request merged**